src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java

```diff
@@ -298,4 +347,166 @@ private static int timeUnitToScale(LogicalTypeAnnotation.TimeUnit timeUnit) {
             ErrorCode.INTERNAL_ERROR, String.format("Unknown time unit: %s", timeUnit));
     }
   }
+
+  /**
+   * Parses a group value based on Parquet group logical type.
+   *
+   * @param value value to parse
+   * @param type Parquet column type
+   * @param statsMap column stats map to update
+   * @param defaultTimezone default timezone to use for timestamp parsing
+   * @param insertRowsCurrIndex Used for logging the row of index given in insertRows API
+   * @param path dot path of the column
+   * @param isdDescendantsOfRepeatingGroup true if the column is a descendant of a repeating group,
+   * @return list of parsed values
+   */
+  private static ParquetBufferValue getGroupValue(
+      Object value,
+      GroupType type,
+      Map<String, RowBufferStats> statsMap,
+      ZoneId defaultTimezone,
+      final long insertRowsCurrIndex,
+      String path,
+      boolean isdDescendantsOfRepeatingGroup) {
+    LogicalTypeAnnotation logicalTypeAnnotation = type.getLogicalTypeAnnotation();
+    if (logicalTypeAnnotation == null) {
+      return getStructValue(
+          value,
+          type,
+          statsMap,
+          defaultTimezone,
+          insertRowsCurrIndex,
+          path,
+          isdDescendantsOfRepeatingGroup);
+    }
+    if (logicalTypeAnnotation instanceof LogicalTypeAnnotation.ListLogicalTypeAnnotation) {
+      return get3LevelListValue(value, type, statsMap, defaultTimezone, insertRowsCurrIndex, path);
+    }
+    if (logicalTypeAnnotation instanceof LogicalTypeAnnotation.MapLogicalTypeAnnotation) {
+      return get3LevelMapValue(
+          value,
+          type,
+          statsMap,
+          defaultTimezone,
+          insertRowsCurrIndex,
+          path,
+          isdDescendantsOfRepeatingGroup);
+    }
+    throw new SFException(
+        ErrorCode.UNKNOWN_DATA_TYPE, logicalTypeAnnotation, type.getClass().getSimpleName());
+  }
+
+  /**
+   * Parses a struct value based on Parquet group logical type. The parsed value is a list of
+   * values, where each element represents a field in the group. For example, an input {@code
+   * {"field1": 1, "field2": 2}} will be parsed as {@code [1, 2]}.
+   */
+  private static ParquetBufferValue getStructValue(
+      Object value,
+      GroupType type,
+      Map<String, RowBufferStats> statsMap,
+      ZoneId defaultTimezone,
+      final long insertRowsCurrIndex,
+      String path,
+      boolean isdDescendantsOfRepeatingGroup) {
+    Map<String, Object> structVal =
+        DataValidationUtil.validateAndParseIcebergStruct(
+            type.getName(), value, insertRowsCurrIndex);
+    List<Object> listVal = new ArrayList<>(type.getFieldCount());
+    float estimatedParquetSize = 0f;
+    for (int i = 0; i < type.getFieldCount(); i++) {
+      ParquetBufferValue parsedValue =
+          parseColumnValueToParquet(
+              structVal.getOrDefault(type.getFieldName(i), null),
```

Comment:
We're missing validation for any unexpected fields being sent in via the map.
Lets do the following:
1. Create a new HashSet(structVal.keySet()) at the beginning of this method
2. For every field that you've validated in this for loop, remove it from this set
3. At the end of the method, the set should be empty. If there's still something left in the set, throw a validation error for unexpected field or whatever the SDK generally throws. With a proper path to the current field of course.

Reply:
Modified.